Firestore/src/FirestoreSessionHandler.php

```diff
@@ -0,0 +1,267 @@
+ * properly handle errors on `write` operations.
+ *
+ * It uses the session.save_path as the Firestore namespace for isolating the
+ * session data from your application data, it also uses the session.name as
+ * the Firestore kind, the session id as the Firestore id. By default, it
+ * does nothing on gc for reducing the cost. Pass positive value up to 1000
+ * for $gcLimit parameter to delete entities in gc.
+ *
+ *
+ * Example without error handling:
+ * ```
+ * use Google\Cloud\Firestore\FirestoreClient;
+ *
+ * $firestore = new FirestoreClient(['projectId' => $projectId]);
+ *
+ * $handler = new FirestoreSessionHandler($firestore);
+ *
+ * session_set_save_handler($handler, true);
+ * session_save_path('sessions');
+ * session_start();
+ *
+ * // Then read and write the $_SESSION array.
+ *
+ * ```
+ *
+ * The above example automatically writes the session data. It's handy, but
+ * the code doesn't stop even if it fails to write the session data, because
+ * the `write` happens when the code exits. If you want to know the session
+ * data is correctly written to the Firestore, you need to call
+ * `session_write_close()` explicitly and then handle `E_USER_WARNING`
+ * properly like the following example.
+ *
+ * Example with error handling:
+ *
+ * ```
+ * use Google\Cloud\Firestore\FirestoreClient;
+ *
+ * $firestore = new FirestoreClient(['projectId' => $projectId]);
+ *
+ * $handler = new FirestoreSessionHandler($firestore);
+ * session_set_save_handler($handler, true);
+ * session_save_path('sessions');
+ * session_start();
+ *
+ * // Then read and write the $_SESSION array.
+ *
+ * function handle_session_error($errNo, $errStr, $errFile, $errLine) {
+ *     # We throw an exception here, but you can do whatever you need.
+ *     throw new Exception("$errStr in $errFile on line $errLine", $errNo);
+ * }
+ * set_error_handler('handle_session_error', E_USER_WARNING);
+ * // If `write` fails for any reason, an exception will be thrown.
+ * session_write_close();
+ * restore_error_handler();
+ * // You can still read the $_SESSION array after closing the session.
+ * ```
+ * @see http://php.net/manual/en/class.sessionhandlerinterface.php SessionHandlerInterface
+ */
+class FirestoreSessionHandler implements SessionHandlerInterface
+{
+    /* @var int */
+    private $gcLimit;
+    /* @var FirestoreClient */
+    private $firestore;
+    /* @var CollectionReference */
+    private $collection;
+
+    /**
+     * Create a custom session handler backed by Cloud Firestore.
+     *
+     * @param FirestoreClient $firestore Firestore client.
+     * @param int $gcLimit [optional] The number of entities to delete in the
+     *        garbage collection. Values larger than 1000 will be limited to
+     *        1000. **Defaults to** `0`, indicating garbage collection is
+     *        disabled by default.
+     *        garbage collection.  Defaults to 0 which means it does nothing.
```

Comment:
sorry, i should've noted this in the review. github suggestions can't span multiple lines, so we'll need to remove this and the next line.